### ISSUE TITLE
Read outside k compilation fails

### ISSUE THREAD

**Issue by tehrengruber** [GridTools/gt4py#582]
The following stencil (adoption of `tests/test_integration/test_suites.py:TestReadOutsideKInterval`) throws an exception during compilation in the dace horizontal merging pass:
```python
def definition(field_in, field_out):
    with computation(PARALLEL), interval(-1, None):
        field_out = (
            field_in[0, 0, -1] + field_in[0, 0, 1]
        )
```

<details>
  <summary>Error message. Click to expand!</summary>
  
```
/home/tille/Development/gt4py/venv/bin/python /usr/share/pycharm/plugins/python-ce/helpers/pydev/pydevd.py --multiproc --qt-support=auto --client 127.0.0.1 --port 36065 --file /usr/share/pycharm/plugins/python-ce/helpers/pycharm/_jb_pytest_runner.py --target tests/test_integration/test_suites.py::TestReadOutsideKInterval.test_generation -- -k "TestReadOutsideKInterval and gtc:dace_field_in_float64_field_out_float64"
Testing started at 2:40 PM ...
Connected to pydev debugger (build 212.5457.59)
Launching pytest with arguments -k TestReadOutsideKInterval and gtc:dace_field_in_float64_field_out_float64 tests/test_integration/test_suites.py::TestReadOutsideKInterval::test_generation --no-header --no-summary -q in /home/tille/Development/gt4py

============================= test session starts ==============================
collecting ... collected 11 items / 10 deselected / 1 selected

tests/test_integration/test_suites.py::TestReadOutsideKInterval::test_generation[gtc:dace_field_in_float64_field_out_float64] <- src/gt4py/testing/suites.py FAILED [100%]Falsifying example: hyp_wrapper(
    hypothesis_data={},
    test_hyp={'backend': 'gtc:dace',
     'constants': {},
     'definition': <function tests.test_integration.test_suites.definition>,
     'dtypes': {'field_in': dtype('float64'), 'field_out': dtype('float64')},
     'generation_strategy': <function gt4py.testing.input_strategies.accept.<locals>.strategy>,
     'implementations': [],
     'marks': (),
     'suite': 'TestReadOutsideKInterval',
     'test_id': 4},
)

src/gt4py/testing/suites.py:199 (TestReadOutsideKInterval.test_generation[gtc:dace_field_in_float64_field_out_float64])
self = <tests.test_integration.test_suites.TestReadOutsideKInterval object at 0x7efe9fcb7fa0>
test = {'backend': 'gtc:dace', 'constants': {}, 'definition': <function definition at 0x7efe9fce6430>, 'dtypes': {'field_in': dtype('float64'), 'field_out': dtype('float64')}, ...}

    def generation_test_wrapper(self, test):
        @hyp.given(hypothesis_data=test["generation_strategy"]())
        def hyp_wrapper(test_hyp, hypothesis_data):
            self._test_generation(
                test_hyp, {**test_hyp["constants"], **hypothesis_data, **cls_dict["singletons"]}
            )
    
>       hyp_wrapper(test)

src/gt4py/testing/suites.py:207: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/gt4py/testing/suites.py:202: in hyp_wrapper
    def hyp_wrapper(test_hyp, hypothesis_data):
src/gt4py/testing/suites.py:203: in hyp_wrapper
    self._test_generation(
src/gt4py/testing/suites.py:447: in _test_generation
    implementation = gtscript.stencil(
src/gt4py/gtscript.py:281: in stencil
    return _decorator(definition)
src/gt4py/gtscript.py:269: in _decorator
    out = gt_loader.gtscript_loader(
src/gt4py/loader.py:72: in gtscript_loader
    stencil_class = load_stencil("gtscript", backend, definition_func, externals, build_options)
src/gt4py/loader.py:58: in load_stencil
    return builder.build()
src/gt4py/stencil_builder.py:87: in build
    stencil_class = self.backend.generate()
src/gt4py/backend/gtc_backend/dace/backend.py:339: in generate
    pyext_module_name, pyext_file_path = self.generate_extension()
src/gt4py/backend/gtc_backend/dace/backend.py:326: in generate_extension
    return self.make_extension(gt_version=2, ir=self.builder.definition_ir, uses_cuda=False)
src/gt4py/backend/gt_backends.py:751: in make_extension
    gt_pyext_sources: Dict[str, Any] = self.make_extension_sources(ir=ir)
src/gt4py/backend/gt_backends.py:794: in make_extension_sources
    gt_pyext_sources = gt_pyext_generator(ir)
src/gt4py/backend/gtc_backend/dace/backend.py:59: in __call__
    oir = oir_pipeline.run(base_oir)
src/gtc/passes/oir_pipeline.py:103: in run
    oir = step(oir)
src/gtc/passes/oir_dace_optimizations/horizontal_execution_merging.py:288: in graph_merge_horizontal_executions
    return optimize_horizontal_executions(node, GraphMerging)
src/gtc/passes/oir_dace_optimizations/api.py:39: in optimize_horizontal_executions
    sdfg = OirSDFGBuilder().visit(stencil)
src/eve/visitors.py:139: in visit
    return visitor(node, **kwargs)
src/gtc/dace/oir_to_dace.py:657: in visit_Stencil
    return StencilOirSDFGBuilder.build(node.name, node, library_nodes)
src/gtc/dace/oir_to_dace.py:381: in build
    res.validate()
venv/lib/python3.8/site-packages/dace/sdfg/sdfg.py:1997: in validate
    validate_sdfg(self)
venv/lib/python3.8/site-packages/dace/sdfg/validation.py:142: in validate_sdfg
    validate_state(start_state, sdfg.node_id(start_state), sdfg,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

state = SDFGState (TestReadOutsideKInterval_gtcdace_4_state), state_id = 0
sdfg = <dace.sdfg.sdfg.SDFG object at 0x7efe9f561400>
symbols = {'__I': int, '__J': int, '__K': int, '__field_in_I_stride': int, ...}

....

>                       raise InvalidSDFGEdgeError("Memlet subset out-of-bounds",
                                                   sdfg, state_id, eid)
E                       dace.sdfg.validation.InvalidSDFGEdgeError: Memlet subset out-of-bounds (at state TestReadOutsideKInterval_gtcdace_4_state, edge field_in[0:__I, 0:__J, __K - 2:__K + 1] (field_in:None -> VerticalLoop_139632060011088:IN_field_in))

venv/lib/python3.8/site-packages/dace/sdfg/validation.py:482: InvalidSDFGEdgeError
======================= 1 failed, 10 deselected in 4.09s =======================

Process finished with exit code 1
```
</details>

Note that for `interval(...)`, `interval(0, 1)`, `interval(1, None)` everything is fine.

__EDIT__: Wait until https://github.com/GridTools/gt4py/issues/568 is sorted out

**Comment by havogt** [comment]
In #596 I summarized the rule that we discussed for legal bounds. According to that, the code should be valid if `field_in` is big enough that `field_in[0,0,1]` is within bounds for all k in the domain.

**Comment by tehrengruber** [comment]
Closed by #597
